### PR TITLE
Switch Llama3 model to use the new model implementation

### PR DESCRIPTION
**Pull request by bzgoogle**
# Description

Switch to use new model implementation for Llama3 models

The results from the legacy and new Llama3 models are shown below:

**MMLU**
* Accuracy:
  - Legacy Llama3-8b: [0.78](https://paste.googleplex.com/5795480268963840#l=52)
  - New Llama3-8b: [0.76](https://paste.googleplex.com/5415776638730240#l=54)
 
* Performance:
  -  xProf on single prompt before final logit decoding :
     -  Legacy Llama3-8b: [3.389 ms](https://screenshot.googleplex.com/PB7QJTBQZDarvhc)
     - New Llama3-8b: [3.114 ms](https://screenshot.googleplex.com/8YKoh4FxxnLuDwL)
  - Throughput:
     -  Legacy Llama3-8b: [41267 tk/s](https://paste.googleplex.com/5795480268963840#l=24)
     - New Llama3-8b: [42712 tk/s](https://paste.googleplex.com/5415776638730240#l=26)
     - Legacy Llama3-70b: [26089 tk/s](https://paste.googleplex.com/4811125438545920#l=22)
     - New Llama3-70b: [25618 tk/x](https://paste.googleplex.com/6468380764405760#l=22)

**MLPerf**
* Accuracy
  - Legacy Llama3-8b: [44.4](https://screenshot.googleplex.com/5Es7ADrWZGCPDmU)
  - New Llama3-8b: [44.3](https://paste.googleplex.com/6037885773676544#l=40)
 
* Performance:
  - Throughput:
     -  Legacy Llama3-8b: [4496 tk/s](https://screenshot.googleplex.com/5Es7ADrWZGCPDmU)
     - New Llama3-8b: [4450 tk/s](https://paste.googleplex.com/6037885773676544#l=43)
   
# Tests

CI: https://buildkite.com/tpu-commons/tpu-commons-ci/builds/764

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by wwl2755-google** [comment]
Has the old design already been deprecated? We may need some time to hold both alive to make sure the transit is smooth.

File: tpu_commons/models/jax/recipes/llama3.py [hunk 1]
```diff
@@ -102,17 +102,19 @@ class Llama3RecipeConfig(RecipeConfig):
...
         self.vllm_config = vllm_config
         self.rng = nnx.Rngs(rng)
         self.mesh = mesh
         try:
             strategy_dict = self.vllm_config.additional_config["sharding"][
                 "sharding_strategy"]
         except (KeyError, TypeError):
-            strategy_dict = {"tensor_parallelism": 4, "expert_parallelism": 2}
+            strategy_dict = {"tensor_parallelism": 1}
+        #TODO: after all models are migrated to the new sharding,
```
Comment (jrplatin) [inline tpu_commons/models/jax/recipes/llama3.py:116]: Is this a relevant TODO?

Comment (bzgoogle) [reply tpu_commons/models/jax/recipes/llama3.py:116]: I think so. we will remove the sharding initialization here after all models are migrated

**Comment by bzgoogle** [comment]
> Has the old design already been deprecated? We may need some time to hold both alive to make sure the transit is smooth.

Hi @wwl2755-google , the existing & new models have been alive for the past weeks. We achieved the accuracy & perf parity right now, and would like to depreciate the old one. What do you referring to for a smooth transit?

**Comment by wwl2755-google** [comment]
> > Has the old design already been deprecated? We may need some time to hold both alive to make sure the transit is smooth.
> 
> Hi @wwl2755-google , the existing & new models have been alive for the past weeks. We achieved the accuracy & perf parity right now, and would like to depreciate the old one. What do you referring to for a smooth transit?

Thanks for your clarification. So you are only deleting the llama. Then it makes sense to me. Just please don't delete the whole path because there is still some ongoing work to be done there. Cheers :-)

**Comment by wwl2755-google** [comment]
>  We achieved the accuracy & perf parity right now, and would like to depreciate the old one. 

Hi @bzgoogle , Thanks so much for your effort on this, but deprecating an existing solution in well-condition could be dangerous so have to be extremely careful. I just checked some recent CIs ([CI1](https://buildkite.com/tpu-commons/tpu-commons-ci/builds/812#01982ac0-fcb6-4c54-8bb4-c057120e0181) , [CI2](https://buildkite.com/tpu-commons/tpu-commons-ci/builds/807#01982a86-5adc-48b3-b03d-1a7406ebea47), and [CI3](https://buildkite.com/tpu-commons/tpu-commons-ci/builds/805#01982a56-2f0a-4a30-90df-b3dca89c4a66)), all of which shows new design is consistently less performed (4200 vs 4600 tok/s).

Could you double-check what's going on before deleting the old stuffs? Or let CI make several more runs to see the real number? Thanks again for your effort and patience.

File: tpu_commons/models/jax/recipes/llama3.py [hunk 2]
```diff
@@ -102,17 +102,19 @@ class Llama3RecipeConfig(RecipeConfig):
     serving: Llama3ServingConfig = field(default_factory=Llama3ServingConfig)
 
 
-class Llama3(Model):
+class LlamaForCausalLM(Model):
 
-    def __init__(self, vllm_config: VllmConfig, rng: PRNGKey, mesh: Mesh):
+    def __init__(self, vllm_config: VllmConfig, rng: jax.Array, mesh: Mesh):
```
Comment (patemotter) [inline tpu_commons/models/jax/recipes/llama3.py:107]: Are we sure this is a Jax Array and not a PRNGKey? Throughout the code we seem to be going back and forth on the type hint between the two.

Comment (bzgoogle) [reply tpu_commons/models/jax/recipes/llama3.py:107]: Thanks for spotting this. It should be jax Array, fill fix it in a follow-up PR for the hint

**Comment by bzgoogle** [comment]
> > We achieved the accuracy & perf parity right now, and would like to depreciate the old one.
> 
> Hi @bzgoogle , Thanks so much for your effort on this, but deprecating an existing solution in well-condition could be dangerous so have to be extremely careful. I just checked some recent CIs ([CI1](https://buildkite.com/tpu-commons/tpu-commons-ci/builds/812#01982ac0-fcb6-4c54-8bb4-c057120e0181) , [CI2](https://buildkite.com/tpu-commons/tpu-commons-ci/builds/807#01982a86-5adc-48b3-b03d-1a7406ebea47), and [CI3](https://buildkite.com/tpu-commons/tpu-commons-ci/builds/805#01982a56-2f0a-4a30-90df-b3dca89c4a66)), all of which shows new design is consistently less performed (4200 vs 4600 tok/s).
> 
> Could you double-check what's going on before deleting the old stuffs? Or let CI make several more runs to see the real number? Thanks again for your effort and patience.

Hi Wenlong, the some gaps maybe due to the scheduler change the number of chucked tokens between the 2 arms, i.e. https://screenshot.googleplex.com/7seVCtzeHibfk2z and https://screenshot.googleplex.com/BNcsrUaUZRWRTnw . So I will merge this PR and investigate those in a separate buganizer
